backend/models/modelSetting.js

```diff
@@ -27,318 +25,499 @@ const PermissionTemplates = require("./permissionTemplates");
+
+	if (!setting) {
+		throw responseCodes.MODEL_INFO_NOT_FOUND;
+	}
+
+	return ModelSetting.populateUsers(account, setting.permissions);
 };
 
 /**
  * Fills out the models data for the  modelids passed through parameter.
- * @param {Object} teamSpaces an object which keys are teamspaces ids and values are an array of modelids
+ * @param {Object} teamspaces an object which keys are teamspaces ids and values are an array of modelids
  * @returns {Object} which contains the models data
   */
-schema.statics.getModelsData = function(teamSpaces) {
+ModelSetting.getModelsData = function(teamspaces) {
 	return Promise.all(
-		Object.keys(teamSpaces).map(accountDB => {
-			const modelsIds = teamSpaces[accountDB];
+		Object.keys(teamspaces).map((account) => {
+			const modelsIds = teamspaces[account];
 
-			return db.getCollection(accountDB, MODELS_COLL)
-				.then(collection => collection.find({_id: {$in:modelsIds}},{ name: 1, federate: 1, _id: 1}).toArray())
-				.then(models => {
+			return db.find(account, MODELS_COLL, {_id: {$in:modelsIds}}, { name: 1, federate: 1, _id: 1})
+				.then((models) => {
 					const res = {};
 					const indexedModels = models.reduce((ac,c) => {
 						const obj = {}; obj[c._id] = c; return Object.assign(ac,obj); // indexing by model._id
 					} ,{});
-					res[accountDB] = indexedModels;
+					res[account] = indexedModels;
+
 					return res;
 				});
 		})
-	).then((modelData)=> modelData.reduce((ac,cur) => Object.assign(ac, cur),{})); // Turns the array to an object (quick indexing);
+	).then((modelData) => modelData.reduce((ac,cur) => Object.assign(ac, cur),{})); // Turns the array to an object (quick indexing);
+};
+
+ModelSetting.isFederation = async function(account, model) {
+	const modelSetting = await ModelSetting.findModelSettingById(account, model);
+
+	if (!modelSetting) {
+		throw responseCodes.MODEL_NOT_FOUND;
+	} else if (!modelSetting.federate) {
+		throw responseCodes.MODEL_IS_NOT_A_FED;
+	}
+
+	return true;
+};
+
+ModelSetting.isModelNameExists = async function(account, models, modelName) {
+	const coll = await db.getCollection(account, MODELS_COLL);
+	const count = await coll.find({name: modelName, _id: {"$in": models}}).count();
+
+	return count > 0;
+};
+
+ModelSetting.populateUsers = async function(account, permissions) {
+	const { getAllUsersInTeamspace } = require("./user");
+
+	const users = await getAllUsersInTeamspace(account);
+
+	users.forEach(user => {
+		const permissionFound = permissions && permissions.find(p => p.user ===  user);
+
+		if (!permissionFound) {
+			permissions.push({ user });
+		}
+	});
+
+	return permissions;
+};
+
+ModelSetting.populateUsersForMultiplePermissions = function (account, permissionsList) {
+	const promises = permissionsList.map((permissions) => {
+		return ModelSetting.populateUsers(account, permissions);
+	});
+
+	return Promise.all(promises);
+};
+
+ModelSetting.setModelImportFail = async function(account, model, errorReason) {
```

Comment:
this (and the few functions that follow this) should really just do update instead of find, manipulate the data, then call the generic update function. (or findOneAndUpdate if you need the settings to be returned)

you don't need validation checks here since there's no user input.

Reply:
@carmenfan For `setModelImportFail` it needs to know if `setting.type` is `toy|sample`

Reply:
I'm not sure why it's wiping the timestamp if the federation failed tbh... that doesn't make a lot of sense

`setModelImportSuccess` is passing in isToy, do you know if there's a reason why it's not doing the same here?